R/saveDeAsJson.R

```diff
@@ -0,0 +1,118 @@
+#' }
+#' ## The results can be viewed in a webbrowser by opening toc.html
+#'
+#' @export
+saveDeAsJson <- function(de.raw, sample.groups=NULL, saveprefix=NULL, dir.name="JSON", gene.metadata=NULL,verbose=TRUE) {
+
+  checkPackageInstalled("jsonlite", cran=TRUE)
+  if (!is.null(dir.name)) {
+    if (!dir.exists(dir.name)) dir.create(dir.name)
+  } else {
+    dir.name <- "."
+  }
+
+  if (is.null(gene.metadata)) {
+    gene.metadata <- lapply(de.raw, function(x) {
+      if (!is.null(x)) {
+        rownames(as.data.frame(x$res))
+      } else {
+        NULL
+      }
+    }) %>%
+      unlist() %>%
+      unique() %>%
+      {data.frame(geneid = .)}
+  } else {
+    if(is.null(gene.metadata$gene.id)) stop("gene.metadata must contain $gene.id field")
+  }
+
+  lapply(sn(de.raw %>% names()), function(ncc) {
+    if(verbose) print(ncc)
+    res.celltype <- de.raw[[ncc]]
+    res.table <- res.celltype$res %>% as.data.frame()
+    res.table$gene <- rownames(res.table)
+    res.table$significant <- res.table$padj < 0.05
+    res.table$log2FoldChange[is.na(res.table$log2FoldChange)] <- 0
+    res.table$rowid <- 1:nrow(res.table)
+
+    all.genes <- rownames(res.table)
+    cm <- res.celltype$cm
+
+    ilev <- lapply(sample.groups, function(sg) {
+      sg <- sg[sg %in% colnames(cm)]
+      cm.tmp <- cm[,sg]
+      cm.tmp <- as.matrix(cm.tmp)
+      rownames(cm.tmp) <- rownames(cm)
+
+      ## calculate cpm
+      cpm <- sweep(cm.tmp, 2, apply(cm.tmp,2, sum), FUN='/')
+      cpm <- log10(cpm * 1e6 + 1)
+      snames1 <- colnames(cpm)
+
+      ## Put genes in order
+      cpm <- cpm[all.genes,]
+      colnames(cpm) <- NULL;
+      rownames(cpm) <- NULL;
+
+      list(snames=snames1, val=as.matrix(cpm))
+    })
+
+    snames <- names(res.celltype$sample.groups)
+
+    ## convert to json
+    tojson <- list(
+      res = res.table,
+      genes = all.genes,
+      ilev = ilev,
+      snames = snames)
+    y <- jsonlite::toJSON(tojson)
+    file <- paste0(dir.name, "/", saveprefix, make.names(ncc), ".json")
+    write(y, file)
+    NULL
+  })
+
+  toc.file <- paste0(dir.name, "/toc.html")
+  s <- c(list('<html><head><style>
+    table {
+    font-family: arial, sans-serif;
+    border-collapse: collapse;
+    width: 100%;
+    }
+
+    td, th {
+    border: 1px solid #dddddd;
+    text-align: left;
+    padding: 8px;
+    }
+
+    tr:nth-child(even) {
+    background-color: #dddddd;
+    }
+
+    </style></head><body><table>'),
+         lapply(names(de.raw), function(n)
+           paste0('<tr><td><a href="deview.2.html?d=', saveprefix, make.names(n),'.json">', n, '</a></td></tr>')),
+         list('</table></body></html>')
+  ) %>% paste(collapse='\n')
+
+  write(s, file=toc.file)
+
+  # Copy viewer files
+  file.copy(from = c(paste(system.file(package="sccore"), "extdata/deview.2.html", sep ="/"),
+                     paste(system.file(package="sccore"), "extdata/deview.2.js", sep ="/")), to = paste0(dir.name), overwrite = TRUE)
```

Comment:
I'd strongly suspect this not to work on Windows because of "/" in the path... In any case, the [recommended way](https://r-pkgs.org/data.html#filepaths) `system.file("extdata", "deview.2.html", package="sccore")` would be shorter and safer.

Reply:
Btw, it can also be vectorised as
```r
system.file("extdata", c("deview.2.html", "deview.2.js"), package="sccore")
```
:)

Reply:
Could we make these `deview2.html` and `deview.2.js` please?

Reply:
Added Viktor's suggestion